config: Overrides for system and user scopes

haampie opened the pull request:
Closes #11919.
Closes #11951.
Closes #13057.
Closes #12892.
Closes #26341.
Closes #23760.
Closes #23596.
Closes #25547.
Closes #26833.

Short story:

> Spack now provides three environment variables that allow you to override or opt out of configuration locations:
> 
> * ``SPACK_USER_CONFIG_PATH``: Override the path to use for the ``user`` (``~/.spack``) scope.
> * ``SPACK_SYSTEM_CONFIG_PATH``: Override the path to use for the ``system`` (``/etc/spack``) scope.
> * ``SPACK_DISABLE_LOCAL_CONFIG``: set this environment variable to a non-empty value to completely disable **both** the system and user configuration directories. Spack will only consider its own defaults and ``site`` configuration locations.
> 
> And one that allows you to move the default cache location:
> 
> * ``SPACK_USER_CACHE_PATH``: Override the default path to use for user data (misc_cache, tests, reports, etc.)
> 
> With these settings, if you want to isolate Spack in a CI environment, you can do this:
> 
>     export SPACK_DISABLE_LOCAL_CONFIG=true
>     export SPACK_USER_CACHE_PATH=/tmp/spack

This PR is the effective minimal and hopefully least controversial changeset to
address common complaints voiced many times on Slack and Github about
hard-coded config paths in Spack without a way to opt out.

Pinging @trws, @healther, @tjfulle, @jw500, @sleak-lbl, @bartlettroscoe, @dremerb,
@huqy, @eugeneswalker, @mathstuf, @mpbelhorn, @mamelara, @marcmengel who
all contributed to issues and prs addressing something along these lines.

It makes it possible for the user to change the default location of
the user config `~/.spack` and system config `/etc/spack` folders.

This is a stop-gap approach until we have figured out how to deal with
the system and user config scopes more generally, as there are plans to
potentially / eventually get rid of them.

Why make this change?

**User config**

Spack is a bit of a pain when you have:

- a shared $HOME folder across different systems.
- multiple Spack versions on the same system.

**System config**

- On shared systems with a versioned programming environment / toolkit,
  system administrators want to provide config for each version (e.g.
  21.09, 21.10) of the programming environment, and the user Spack
  instance should be able to pick this up without a steep learning
  curve.
- On shared systems the user should be able to opt out of the
  hard-coded config scope in /etc/spack, since it may be incompatible with their
  particular instance. Currently Spack can only opt out of all config
  scopes through overrides with `"config:":`, `"packages:":`, but that
  also drops the defaults config, which would have to be repeated, which
  is undesirable, especially the lengthy packages.yaml.

An example use case is: having config in this folder:

```
/path/to/programming/environment/{version}/{compilers,packages}.yaml
```

and have `module load spack-system-config` set the variable

```
SPACK_SYSTEM_CONFIG_PATH=/path/to/programming/environment/{version}
```

where the user no longer has to worry about what `{version}` they are
on.

**Continuous integration**

Finally, there is the use case of continuous integration, which may
clone an arbitrary Spack version, which optimally should not pick up
system or user config from the previous run (like may happen in
classical bare metal non-containerized filesystem stateful
jenkins pipelines). In fact this is very similar to how spack itself
tries to avoid picking up system dependencies during builds...

**But environments solve this?**

- You could do `include`s in environment files to get similar behavior
  to the spack_system_config_path example, but environments
  1) require you to list paths to individual config files, not directories;
  2) fail if the listed config file does not exist.
- They allow you to override config scopes, but this is generally too
  rigorous, as it requires you to repeat the default config, in
  particular packages.yaml, and just defies the point of layered config.

haampie committed:
config: overrides for caches and system and user scopes (#26735)

Spack's `system` and `user` scopes provide ways for administrators and
users to set global defaults for all Spack instances, but for use cases
where one wants a clean Spack installation, these scopes can be undesirable.
For example, users may want to opt out of global system configuration, or
they may want to ignore their own home directory settings when running in
a continuous integration environment.

Spack also, by default, keeps various caches and user data in `~/.spack`,
but users may want to override these locations.

Spack provides three environment variables that allow you to override or
opt out of configuration locations:

 * `SPACK_USER_CONFIG_PATH`: Override the path to use for the
   `user` (`~/.spack`) scope.

 * `SPACK_SYSTEM_CONFIG_PATH`: Override the path to use for the
   `system` (`/etc/spack`) scope.

 * `SPACK_DISABLE_LOCAL_CONFIG`: set this environment variable to completely
   disable *both* the system and user configuration directories. Spack will
   only consider its own defaults and `site` configuration locations.

And one that allows you to move the default cache location:

 * `SPACK_USER_CACHE_PATH`: Override the default path to use for user data
   (misc_cache, tests, reports, etc.)

With these settings, if you want to isolate Spack in a CI environment, you can do this:

   export SPACK_DISABLE_LOCAL_CONFIG=true
   export SPACK_USER_CACHE_PATH=/tmp/spack

This is a stop-gap approach until we have figured out how to deal with
the system and user config scopes more generally, as there are plans to
potentially / eventually get rid of them.

**User config**

Spack is a bit of a pain when you have:

- a shared $HOME folder across different systems.
- multiple Spack versions on the same system.

**System config**

- On shared systems with a versioned programming environment / toolkit,
  system administrators want to provide config for each version (e.g.
  21.09, 21.10) of the programming environment, and the user Spack
  instance should be able to pick this up without a steep learning
  curve.
- On shared systems the user should be able to opt out of the
  hard-coded config scope in /etc/spack, since it may be incompatible
  with their particular instance. Currently Spack can only opt out of all
  config scopes through overrides with `"config:":`, `"packages:":`, but that
  also drops the defaults config, which would have to be repeated, which
  is undesirable, especially the lengthy packages.yaml.

An example use case is: having config in this folder:

```
/path/to/programming/environment/{version}/{compilers,packages}.yaml
```

and have `module load spack-system-config` set the variable

```
SPACK_SYSTEM_CONFIG_PATH=/path/to/programming/environment/{version}
```

where the user no longer has to worry about what `{version}` they are
on.

**Continuous integration**

Finally, there is the use case of continuous integration, which may
clone an arbitrary Spack version, which optimally should not pick up
system or user config from the previous run (like may happen in
classical bare metal non-containerized filesystem side effect ridden
jenkins pipelines). In fact this is very similar to how spack itself
tries to avoid picking up system dependencies during builds...

**But environments solve this?**

- You could do `include`s in environment files to get similar behavior
  to the spack_system_config_path example, but environments require you
  to:
  1) require paths to individual config files, not directories.
  2) fail if the listed config file does not exist
- They allow you to override config scopes, but this is generally too
  rigurous, as it requires you to repeat the default config, in
  particular packages.yaml, and just defies the point of layered config.

Co-authored-by: Tom Scogland <tscogland@llnl.gov>
Co-authored-by: Tim Fuller <tjfulle@sandia.gov>
Co-authored-by: Steve Leak <sleak@lbl.gov>
Co-authored-by: Todd Gamblin <tgamblin@llnl.gov>

trws commented:
FWIW, I'm all for this.  Admittedly I would prefer that it go further, especially making it possible to make a spack self-contained with a file on disk, but any step in this direction seems like a good one to me.

trws commented on lib/spack/spack/paths.py:
```diff
@@ -50,8 +50,12 @@
+
 #: User configuration location
-user_config_path = os.path.expanduser('~/.spack')
+user_config_path = _get_user_config_path()
```
Why this dance rather than:

```python
default_user_config_path = os.path.expanduser('~/.spack')
 
 #: User configuration location
 user_config_path = os.getenv('SPACK_USER_CONFIG_PATH') or default_user_config_path
```

haampie replied:
Wasn't sure how to write a test for it if it was just globals and no functions

trws replied:
I see what you mean.  Normally I would want it to be an init-test, something that actually runs the file then checks it both ways.  Alternately, possibly make it an `_init_paths` function that gets called there and sets the globals? That would leave it testable, but not have a nullary lambda there with just an or in it.  I don't see an actual issue with your implementation, but I could see that getting tripped over in code review pretty easily.

haampie replied:
Generally I agree, an init function would be nice, but I'd rather leave that to some post 0.17.0 PR that does a proper overhaul of globals in Spack :sweat_smile:

If I would write an init function, the least disruptive one would be one that sets all paths globals. But then if I run it in tests the tests have side effects. And that requires cleanup utilities. And that would make the PR more work to review.

Right now the tests are close to useless (unit testing a lambda function...), but at least they don't have side-effects.

trws replied:
Fair point, it can wait. 👍

trws reviewed (dismissed):
I would have follow-on PRs to this, but have no issue with anything here.

Option to support XDG and non-sysv system config path setups for example would be nice, but with this we can emulate all of that in the meantime.

@becker33, @alalazo, @tgamblin?

tjfulle commented:
I've been submitting pull requests for several years about this issue and would be happy to see something done about it.  I haven't understood the reluctance of the Spack team to make the user config path overridable by an environment variable - this is a defector standard way of customizing applications in the Linux/HPC worlds.  I've got two comments on this specific PR:

- this seems overly complicated.  Why not just do
 
   ```python
   user_config_path = os.getenv('SPACK_USER_CONFIG_PATH', os.path.expanduser('~/'))
   ```

   This can be simply tested with the following test:

   ```python
   import os
   from spack.util.executable import Executable
   import spack.paths
   def test_user_config_path():
       spack_script = os.path.join(spack.paths.bin_path, 'spack')
       spack_python = Executable('{0} python'.format(spack_script))
       env = os.environ.copy()
       user_config_path ='foo/baz'
       env['SPACK_USER_CONFIG_PATH'] = user_config_path
       spack_python(
           '-c',
           'import spack.paths; assert spack.paths.user_config_path == {0!r}'.format(user_config_path),
           allow_failure=True,
           env=env,
       )
       assert spack_python.returncode == 0
   ```

- this PR does not go far enough.  It needs to replace occurrences of `~/.spack` in `$spack/etc/spack/defaults/config.yaml` with `$user_config_path`.  @trws has an alternative PR that does this.  Without this, spack will still write to `./spack`

haampie commented:
> for several years

> this PR does not go far enough.

Please read the first paragraph, the PR is simple to get it in, hopefully before 0.17 branches off, for which the window is narrowing.

Adding `$user_config_path` adds a new feature people will start to rely on. That may reduce chances of getting this in, since there are plans to drop the config scopes in the future (and I don't want to get into that in this PR). Note that the only relevant config setting you're talking about is `misc_cache: ~/.spack/cache`. After you've moved your user config to `$SPACK_USER_CONFIG_PATH/config.yaml`, just set the `config:misc_cache` value there.

> Why not just do

See the [reply above](https://github.com/spack/spack/pull/26735#discussion_r733465535), there are plans to gradually get rid of all the spack globals and put them in some config/context object, which allows us to do unit testing of its initialization without side effects. Your test works too and is side effect free but starting a python subprocess was again something I feared would spark discussion blocking the actual contribution. Unit testing `_get_system_config_path` at least resembles the future unit test for default settings in spack.

tjfulle commented:
Thanks for the reply @haampie - I'll just register my opinion that I think pulling in `SPACK_USER_CONFIG_PATH` without the corresponding change to the default `misc_cache` is a step backwards.  I have a fork of Spack that I administer that applied a similar patch to your PR.  User's were **very** angry that Spack kept trying to write to `~/.spack` when they set `SPACK_USER_CONFIG_PATH` to something else.  These user's were happy with the default configuration and did not want to write a `config.yaml` just to get spack to write the misc. cache to `user_config_path`.  After applying a patch to write misc. cache where it should be (`user_config_path`) the complaints stopped.

I still maintain this fork of spack and many dozen users at Sandia use it - once this issue is resolved we will all happily ditch the fork and use the mainline spack.

haampie commented:
Calling it a step back is not helpful if you really want something along these lines to be merged. And it really is not, cause: (a) it's enabling you to *opt out* of `~/.spack` as a source where config is read from, which was previously impossible. And (b) you can give better defaults to your angry users by setting `config:misc_cache:[some path]` e.g. `~/.spack/cluster-name/cache`, `$scratch/spack-cache`, or `$SPACK_USER_CONFIG_PATH/.spack/cache` in `$SPACK_SYSTEM_CONFIG/config.yaml`.

For further discussion please resort to https://github.com/spack/spack/issues/12892.

tgamblin commented:
Ok I read through this and it all looks good to me. I made some changes on top of what @haampie has already done.  Specifically, there are now *three* possible overrides:

* `SPACK_USER_CONFIG_PATH`: Override the path to use for the ``user`` (``~/.spack``) scope. Caches that normally reside in this location will also be overridden.
* `SPACK_SYSTEM_CONFIG_PATH`: Override the path to use for the ``system`` (``/etc``) scope.
* `SPACK_DISABLE_LOCAL_CONFIG`: set this environment variable to a non-empty value to disable **both** the system and user configuration directories. Spack will only consider its own defaults and ``site`` configuration locations. An empty temporary directory will be used for local caches.

So in the simple case, you can just set `SPACK_DISABLE_LOCAL_CONFIG=TRUE` and be done with all of this.  It actually *removes* the scopes from configuration rather than just overriding them.  For the user config scope, it *does* make a temporary directory so that things like caches can still work.  The temp dir is cleaned up on exit.  If you need to override either system or user config, you can do that with `SPACK_SYSTEM_CONFIG_PATH` and `SPACK_USER_CONFIG_PATH`.

In addition to this, I brought in @trws's `$user_config` variable from #23760, so that we don't run into the issue that @tjfulle mentioned (that `~/.spack` is hardcoded in `etc/spack/defaults/config.yaml`). I also refactored `spack.paths` so that the places we write to/read from are more apparent (the different directories have their own sections now).

IMO this is ready to go but it probably needs one more look from @haampie.  I think this will get everyone the stopgap they want.

haampie commented on lib/spack/spack/paths.py:
```diff
@@ -41,30 +44,71 @@
+#
+def _get_system_config_path():
+    return os.path.expanduser(
+        os.getenv('SPACK_SYSTEM_CONFIG_PATH') or "/etc"
```
```suggestion
        os.getenv('SPACK_SYSTEM_CONFIG_PATH') or "/etc/spack"
```

haampie commented on lib/spack/spack/test/config.py:
```diff
@@ -1167,3 +1174,57 @@ def test_internal_config_scope_cache_clearing():
+
+def test_system_config_path_is_default_when_env_var_is_empty(working_env):
+    os.environ['SPACK_SYSTEM_CONFIG_PATH'] = ''
+    assert "/etc" == spack.paths._get_system_config_path()
```
```suggestion
    assert "/etc/spack" == spack.paths._get_system_config_path()
```

haampie commented on lib/spack/spack/paths.py:
```diff
@@ -41,30 +44,71 @@
+# There are three environment variables you can use to isolate spack from
+# the host environment:
+# - `SPACK_USER_CONFIG_PATH`: override `~/.spack` location (for config and caches)
+# - `SPACK_SYSTEM_CONFIG_PATH`: override `/etc` configuration scope
```
```suggestion
# - `SPACK_SYSTEM_CONFIG_PATH`: override `/etc/spack` configuration scope
```

haampie commented on lib/spack/docs/configuration.rst:
```diff
@@ -562,3 +565,27 @@ built in and are not overridden by a configuration file. The
+* ``SPACK_USER_CONFIG_PATH``: Override the path to use for the ``user`` (``~/.spack``)
+  scope. Caches that normally reside in this location will also be overridden.
+* ``SPACK_SYSTEM_CONFIG_PATH``: Override the path to use for the ``system`` (``/etc``)
+  scope.
```
```suggestion
* ``SPACK_SYSTEM_CONFIG_PATH``: Override the path to use for the ``system``
  (``/etc/spack``) scope.
```

haampie commented on lib/spack/docs/configuration.rst:
```diff
@@ -562,3 +565,27 @@ built in and are not overridden by a configuration file. The
+
+Spack's ``system`` and ``user`` scopes provide ways for administrators and users to set
+global defaults for all Spack instances, but for use cases where one wants a clean Spack
+installation, these scopes can cause unexpeected behavior. For example, users may want
```
```suggestion
installation, these scopes can cause unexpected behavior. For example, users may want
```

haampie commented on lib/spack/spack/caches.py:
```diff
@@ -25,7 +25,7 @@ def misc_cache_location():
     path = spack.config.get('config:misc_cache')
     if not path:
-        path = os.path.join(spack.paths.user_config_path, 'cache')
+        path = spack.paths.default_misc_cache_path
```
I can't seem to make a suggestion here in github, but maybe

```python
path = spack.config.get('config:misc_cache', spack.paths.default_misc_cache_path)
```

haampie commented on lib/spack/spack/config.py:
```diff
@@ -808,8 +790,35 @@ def _config():
+
+    # System configuration is per machine.
+    # This is disabled if user asks for no local configuration.
+    if not os.getenv("SPACK_DISABLE_LOCAL_CONFIG"):
```
```suggestion
    if "SPACK_DISABLE_LOCAL_CONFIG" not in os.environ:
```

Forcing the user to provide a value suggests that the user should pass a truthy value and then by extension falsy values `SPACK_DISABLE_LOCAL_CONFIG=0` may turn the feature off, which is not the case. So, maybe check for existence instead? This is different from SPACK_*_CONFIG_PATH since we care about the value there, and an empty path does not make sense.

tgamblin replied:
Would you rather support 0, false, and empty as possible false values? Seems cleaner to me.

haampie replied:
Hm, I think set/unset is quite a clear boolean, whereas it's unclear what strings are false-y, so my preference is to just stick to set/unset :thinking: we currently have SPACK_COLOR=always/never/auto which mirrors command line arg values, and SPACK_STACKTRACE which also maps to a command line argument and uses set/unset, not truthy/falsy strings

haampie commented:
@tgamblin I think you're forgetting about the `bootstrap:root:~/.spack/bootstrap` default. That makes me wonder how useful `SPACK_DISABLE_LOCAL_CONFIG` is, since the user would still have to set bootstrap:root to some persistent location to avoid bootstrapping every invocation. So that would have to happen on the spack instance level or on the command line.

Also compiler detection & provider cache are done every time. I think in CI I would just do say

```
SPACK_SYSTEM_CONFIG_PATH="$WORKSPACE/system"
SPACK_USER_CONFIG_PATH="$WORKSPACE/user"
```

so that I have something persistent for a "session", and bootstrapping, compiler detection and caching happens once.

haampie commented:
I've incorporated my own comments, squashed the commits, and added co-authors to the commit message.

I don't want to block this PR much longer, but what remains is my concern above that `SPACK_DISABLE_LOCAL_CONFIG` is not useful enough to make sense, since a temporary home folder makes sense for a session instead of a command invocation, and in that case you're probably better off with `export SPACK_USER_CONFIG_PATH="$(mktemp -d)"` followed by multiple `spack` calls.

Also, on my hpc cluster, bootstrapping is *not* fast :( so it's not really a cost you can afford:

```
$ time SPACK_DISABLE_LOCAL_CONFIG= ./spack/bin/spack spec zlib
Input spec
--------------------------------
zlib

Concretized
--------------------------------
==> Bootstrapping clingo from pre-built binaries
zlib@1.2.11%gcc@10.3.0+optimize+pic+shared arch=cray-cnl7-haswell


real	1m10.741s
user	0m47.624s
sys	0m7.502s
```

alalazo reviewed (commented):
I have the same concern as @haampie but other than that LGTM. Maybe if we unset the local config scopes we can automatically point the bootstrapping store to a fix directory like  `$spack/path/opt/bootstrap`?

haampie commented:
But then where do other caches go? Provider cache is quite time consuming, probably slower than bootstrapping itself. My preference would be to make the user decide where stuff should go, instead of resorting to $tempdir or $spack

tjfulle commented:
I concur with @haampie RE `SPACK_DISABLE_LOCAL_CONFIG`.

trws commented:
Wow, a good bit to unpack here.   I like it!

As to DISABLE, I like the idea, but I'm not sure I like all the details.  As is it would be a useful mode for a one-shot throw-away spack, but not for an instance that needs to stay around.  It is however about 90% of the way to something I personally really want though, which is a way to say "store everything for this spack in the spack root." I'm thinking if we add one more variable, (yeah, I know), to specify the default cache location then we can deal with this.

Pulling from XDG, conceptually one frequently wants each of: data, config, cache and runtime directories (some others too, but they really don't matter here).  No matter what we want the system runtime directories, so that's fine.  We have a way to specify where the stage goes, and it defaults to temp which seems sane, we're adding a setting for configuration, lets add one for cache.  That way we can actually support XDG with some environment variables, which would be nice, *and* if one sets the `SPACK_CACHE_DIR` to `$spack/opt/cache` plus the disable environment variable they get a hermetic spack root by default (except stage and runtime).  It would also allow one of my favorite horrifyingly awful but useful tricks for swap files, set `SPACK_CACHE_DIR` to say `~/.cache/spack-base` and everything else to `$user_cache/$spack/<suffix>` to disambiguate separate spacks but still keep caches together for easy annihilation! 😁

Thoughts?

tgamblin commented:
So you have that, that’s `SPACK_USER_CONFIG_PATH`.  Seems like the issue is really that local config and caches are conflated. I’m tempted to add another env override for `SPACK_USER_CACHE_PATH`, so you can set that to a specific path *and* disable local config.  Would that work?  It’s a little hacky because we already allow overriding this via config, but home directory caches seem special enough to allow something.  

The issue is really that there isn’t another great place for caches. The spack installation is intended to be read-only eventually. Temp and home are really the only places the user is guaranteed to be able to write.

I think eventually we should only need the config disablement, as the reason we want to move the cache right now is because misc cache doesn’t play nice across multiple spack instances. Bootstrap cache works fine AFAIK. I think the thing to do is to add a separate variable for caches, restrict the three variables here already to config, and eventually fix caches so we do not need the cache disablement.

Thoughts?

trws commented:
I forgot about bootstrap... Technically that should go in a more persistent non-cache location I suppose, not sure if that has to have an arch specific base.  If so, that should probably be configurable as well, but we would always do that with a config file. Since the config system can take a ranked list of config locations anyway, it might be worth allowing comma separated SPACK_CONFIG_DIRS like XDG_CONFIG_DIRS to allow layering, but that's gravy if we can do the rest.

tgamblin commented:
Bootstrap does not have to be arch specific. It’s a regular spack install tree.

tgamblin commented:
The other option would be to just fix the misc cache to work across instances. I am not sure I can pull that off before 0.17, but it might be doable.

EDIT: I think it’s probably not since the misc cache really needs git-like size/stat checking as well as an instance identifier of some sort in its path.  I think that’s post-0.17 stuff. We could do it for 0.17.1 and add the env var for 0.17.

trws commented:
I like the idea of being able to place the cache.  The question might be more whether the source and misc cache should go to the same place.

tjfulle commented:
@tgamblin - given how many separate pull requests and issues have been opened for this issue, I believe merging what you currently have (with cache going to `$user_config_path/cache`) is preferable to pushing off till after 0.17.  As you said, the cache location can always be changed in the user's `config.yaml` if they don't like the default.

haampie commented:
> Bootstrap cache works fine AFAIK

The binaries probably work, but otherwise it stores `~/.spack/bootstrap/config/linux/{compilers,compilers,bootstrap}.yaml`. Probably @alalazo can better judge if that is problematic across systems. I'm not sure how these config files even work, given that the location to them is stored in config. Are these config files just stored for posterity like a log (this was what was used during bootstrapping) or are they actually used at runtime?

Btw @tjfulle, last bike shedding discussion is probably whether we should use $user_config_path or $user_config in the config.yaml files :sweat_smile: seems like you automatically write the former but the latter is in this pr.

trws commented:
I thought those went into the user config, compilers.yaml IMO should really be at least spack instance specific.

tjfulle commented:
@haampie - Just muscle memory.  `user_config` is in the PR and I'd like to see it merged sooner than later, so I'll retrain my fingers to type `user_config` 😄

alalazo commented:
> Are these config files just stored for posterity like a log (this was what was used during bootstrapping) or are they actually used at runtime?

They are used during bootstrapping, to be "isolated" from user config so that we can safely perform operations like search for externals and modify configuration without the user being affected. Configuration values that specify general settings for Spack (i.e. everything in `config.yaml`) are copied over.

haampie commented:
@tgamblin yeah, seems like the main issue is that a bunch of things got put in what we now call the user config dir.

But I think people think in terms of the `~/.spack` dir, not in terms of user config dir / cache dir as separate concepts. Given that
1. we sometimes have configurable cache dir paths in user config, and sometimes hard-code them as subdirs to user_config_path in paths.py
2. not everything seems to qualify as cache or config (~/.spack/bootstrap? ~/.spack/reports? module files? ~/.spack/git_repos?)

I'm not sure whether it's a good idea to split those things as part of this pr. As I said, I can live without SPACK_DISABLE_LOCAL_CONFIG as long as SPACK_*_CONFIG_PATH exists, and I would be fine with "abusing" the user config dir by filling it with other stuff

tgamblin commented:
Ok, based on the above, here is a proposal.  If people are ok with this I'll do it:

* `SPACK_USER_CONFIG_PATH`: Override the path to use for the `user` (`~/.spack`) scope. Caches that normally reside in this location will also be overridden.
* `SPACK_SYSTEM_CONFIG_PATH`: Override the path to use for the ``system`` (``/etc``) scope.
* `SPACK_DISABLE_LOCAL_CONFIG`: set this environment variable to a non-empty value to disable **both** the system and user configuration directories. Spack will only consider its own defaults and ``site`` configuration locations.
* `SPACK_MISC_CACHE_PATH` : Override the default location for the misc cache. Normally this is `~/.spack/cache` .  Set to `"tmpdir"` if you want Spack to pick a temporary directory for the misc_cache.

Also, in `config.yaml`:
* `user_config` -> `user_config_path`
* add `misc_cache_path` and make it the default location of `misc_cache`.

Basically this separates concerns about the `misc_cache` from concerns about config.  Notably, it also *omits* the default bootstrap cache location.  If you want to change that, or the default location for tests, you can change it in a config file or in your environment.

Does that work for folks?

I think what this means is that most people will just use `SPACK_DISABLE_LOCAL_CONFIG`.  People who are worried about `misc_cache` can use `SPACK_MISC_CACHE_PATH` until we fix it, and maybe we'll deprecate that one day if things are stable.

mathstuf commented:
> I think what this means is that most people will just use SPACK_DISABLE_LOCAL_CONFIG.

Will that work for the case where I want `spack compiler find` to be able to write its discoveries somewhere?

haampie commented:
I'm genuinely confused why we are heading in a direction where the stop gap solution for 0.17 is more involved than any proposal yet...

We have `config:misc_cache` as a config variable. Is your suggestion to default it to `$misc_cache_path` which is a influenced by an environment variable `SPACK_MISC_CACHE_PATH`?

trws commented:
I think you're right about user config accreting many things that are not config.  Perhaps this should be in phases, and do the full XDG-style split-out, but in order of preference I'd take:

1. This + an option for cache base, 
2. This as is with a follow-on PR
3. Waiting to figure out where everything goes

trws commented:
Sounds great, as long as SPACK_MISC_CACHE_PATH goes through expansion I'm delighted with that @tgamblin.

tgamblin commented:
@mathstuf:

> Will that work for the case where I want spack compiler find to be able to write its discoveries somewhere?

Yes, *if* your spack instance is writable, which it likely is in CI.  Without user config the highest precedence scope will be the poorly named (sorry) "site" scope -- which is in etc/spack in the repo.

If you need more than that you can use an environment or `SPACK_USER_CONFIG_PATH`.

tgamblin commented:
@haampie:

> We have config:misc_cache as a config variable. Is your suggestion to default it to $misc_cache_path which is a influenced by an environment variable `SPACK_MISC_CACHE_PATH`?

Yeah, it's ugly.  Alternately, we could just omit `SPACK_MISC_CACHE_PATH` and leave it to config.  I think that might be cleaner, and it avoids the extra config option.

So, new proposal:

* `SPACK_USER_CONFIG_PATH`: Override the path to use for the `user` (`~/.spack`) scope. Caches that normally reside in this location will also be overridden.
* `SPACK_SYSTEM_CONFIG_PATH`: Override the path to use for the ``system`` (``/etc``) scope.
* `SPACK_DISABLE_LOCAL_CONFIG`: set this environment variable to a non-empty value to disable **both** the system and user configuration directories. Spack will only consider its own defaults and ``site`` configuration locations.

Also, in `config.yaml`:
* `user_config` -> `user_config_path`

If you want to customize `misc_cache`, you either:

1. Set it in your `spack.yaml` in your environment (e.g., for CI)
2. Set it in `etc/spack/config.yaml` in your repo
3. Set `SPACK_USER_CONFIG_PATH` to your own user config location and set `misc_cache` in there.

Does that work for @trws and @tjfulle as well?  It avoids extra complexity and inconsistency in how the caches are handled.

trws commented:
I can work with that.  It seems unfortunate that there's no way to get it to put the caches into the spack root as well with the disable option, but setting the user config path to a path with a yaml file that sets the paths all in terms of `$spack` seems at least possible.  Would you consider something like a particular value for that variable that would redirect caches into the spack directory as well?

tjfulle commented:
@tgamblin I like your last proposal.  In this paradigm, `misc_cache` will default to `user_config_path/cache` and user's can set `misc_cache` in any number of other ways if they want.  Is this right?

My two cents on `SPACK_DISABLE_LOCAL_CONFIG` - the name does not convey what it actually does.  It does not disable the config, just writes it to a mystery location.  It also does not just disable the local config but also the system config.  Why not ditch `SPACK_DISABLE_LOCAL_CONFIG` and allow something like `SPACK_USER_CONFIG_PATH=/dev/null`?  It seems totally unambiguous that the user does not want any configuration written/read.  They can set `SPACK_SYSTEM_CONFIG_PATH` similarly if they want.

tjfulle commented:
@trws - couldn't you do:

```
$ cat $SPACK_USER_CONFIG_PATH/config.yaml | grep misc_cache
  misc_cache: $spack/var/spack/misc_cache
```

Maybe that is what you implied?

tgamblin commented:
> My two cents on `SPACK_DISABLE_LOCAL_CONFIG` - the name does not convey what it actually does. It does not disable the config, just writes it to a mystery location.

This is actually not true.  It does disable user config -- it gets rid of those scopes entirely.  It also creates a temporary directory for *some* of the user caches, because the user config directory is conflated with the location we set for caches.  When that happens there is actually no user config scope.   It's confusing, I agree.

Do people like the `/dev/null` thing better than `SPACK_DISABLE_LOCAL_CONFIG`?  Would you rather have to set that for *both* system and user config?  The intent with `SPACK_DISABLE_LOCAL_CONFIG` was to have something really simple to tell people to isolate their Spack instance, rather than requiring *two* environment variables to do that.

tjfulle commented:
@tgamblin - thanks for the clarification.  I had not realized the scope was completely skipped, my reading of the PR said the scope was still there, but it was in a unnamed temporary directory.  Under this paradigm, where would compiler's be written to?  I presume the `platform` scope would also be disabled since it is relative to the user scope, right?

My preferred arrangement:

- system config scope must be opted in.  I didn't even realize it existed until reading this PR and I don't much like that some system administrator could have say in how I build my software.  Yikes!  Maybe something like
   ```python
   system_config_path = "/etc/spack" if os.getenv("SPACK_USE_SYSTEM_CONFIG") else os.devnull
   ```
- `SPACK_USER_CONFIG_PATH` overrides `~/.spack`.  I set to `/dev/null` if I really want it to be disabled.

with this arrangement, setting `SPACK_USER_CONFIG_PATH=/dev/null` completely isolates Spack.  If a user wants to use a system setting, they can additionally *opt in* to that.

tgamblin commented:
> system config scope must be opted in. I didn't even realize it existed until reading this PR and I don't much like that some system administrator could have say in how I build my software. Yikes! 

We're not getting rid of that by default yet -- it's been there forever.  I think we can deprecate it, but right now I think opting out is better.

> I presume the platform scope would also be disabled since it is relative to the user scope, right?

yep

trws commented:
> @trws - couldn't you do:
> 
> ```
> $ cat $SPACK_USER_CONFIG_PATH/config.yaml | grep misc_cache
>   misc_cache: $spack/var/spack/misc_cache
> ```
> 
> Maybe that is what you implied?

Yeah, that's basically what I meant, but the issue is I need to work with several spacks and would like them to have separate configurations but all have that setting.  It would be nice to be able to get it in one, but I can add a common config to SPACK_CONFIG_PATH or something to get it I think.

tgamblin commented:
Ok folks, trying again.  I *think* this:

1. Makes things simple enough for @haampie.
2. Makes things consistent enough for @haampie.
3. Provides control over local caches *as well as* config.
4. Keeps things simple for most users who want to isolate things in CI.
5. Removes the magic tmpdir stuff that was previously there.

The new new new proposal is (see also docs in the PR):

> Spack provides three environment variables that allow you to override or opt out of configuration locations:
> 
> * ``SPACK_USER_CONFIG_PATH``: Override the path to use for the ``user`` (``~/.spack``) scope.
> * ``SPACK_SYSTEM_CONFIG_PATH``: Override the path to use for the ``system`` (``/etc/spack``) scope.
> * ``SPACK_DISABLE_LOCAL_CONFIG``: set this environment variable to a non-empty value to completely disable **both** the system and user configuration directories. Spack will only consider its own defaults and ``site`` configuration locations.
> 
> And one that allows you to move the default cache location:
> 
> * ``SPACK_USER_CACHE_PATH``: Override the default path to use for user data (misc_cache, tests, reports, etc.)
> 
> With these settings, if you want to isolate Spack in a CI environment, you can do this:
> 
>     export SPACK_DISABLE_LOCAL_CONFIG=true
>     export SPACK_USER_CACHE_PATH=/tmp/spack

Per one of @haampie's observations, I *also* reworked `paths.py` to separate user config from user data.  So, now, bootstrap store, misc cache, reports, test output, `spack monitor` output, etc. are grouped together, and config is an entirely separate thing.  It just *so happens* that the user config scope by default points at `~/.spack`.

What I am not fixing here is that some of these cache locations are hard-coded in `paths.py` and some are overridable in config.  `SPACK_USER_CACHE_PATH` controls a `user_cache_path` variable that is accessible in `config.yaml`, just like `user_config_path`.  You *can* currently override the bootstrap store location and the `misc_cache` location in your config file, and it'll take precedence over `SPACK_USER_CACHE_PATH`.

IMO that is not ideal, and we should in some future release consolidate settings for user directory stuff into a smaller number of settings.  But that's not done here.

Not done here are:
* @tjfulle's `/dev/null` thing but I'm not opposed to adding that later.  Spack will currently likely complain if you set a config scope path to `/dev/null`, but `SPACK_DISABLE_LOCAL_CONFIG` lets you get rid of *all* the host-specific config in one place.  I expect MOST people will just set that to `true` and move on.
* Moving GPG and source cache stuff out of `/var/spack`.  This needs a bit more work.

I think to make this clean, we needed to deconflate the caches in `~/.spack` from the config scope.  There is more reorganization of Spack's install layout that should be done but I think this is what's necessary to get what you all want for the config scope *and* for the caches.

The reason this took so long (and why it is still in flux) is because ultimately I would *really* like to make the spack install prefix read-only and do some sort of reorg like #11871.  We just haven't had time to get to that with all the new concretizer stuff, and I was hoping to *not* have to do a stopgap here because of all the things that are currently wrapped up in `~/.spack`.  Hopefully this is enough for now.

Thoughts?  Trying extremely hard to get to consensus here.

haampie commented on etc/spack/defaults/bootstrap.yaml:
```diff
@@ -4,7 +4,7 @@ bootstrap:
   # Root directory for bootstrapping work. The software bootstrapped
   # by Spack is installed in a "store" subfolder of this root directory
-  root: ~/.spack/bootstrap
+  root: $user_config/bootstrap
```
```suggestion
  root: $user_config_path/bootstrap
```

haampie commented:
Right now `$user_cache_path` is still a bit awkward cause the misc cache is in `$user_cache_path/cache`

tgamblin commented:
> Right now `$user_cache_path` is still a bit awkward cause the misc cache is in `$user_cache_path/cache`

Is this the worst thing?  Changing it will mean that most users end up with `cache` and `misc_cache` in their home directory.  Also `misc_cache` should really be something else in the future.

haampie commented:
Not the worst, but if I think about the cache dir I think about `~/.spack/cache`, not `~/.spack`. Should we move `~/.spack/{reports,git_repos,bootstrap}` under `~/.spack/cache` or leave it?

tgamblin commented:
I'd say leave it for now, as like I said I would like to do another larger rework of where everything lives.  I think we could tackle that stuff then.

haampie commented:
Sure, it's not a big deal

tgamblin commented:
Ok @trws @tjfulle if y'all are good with the solution as described above I think we might have a winner.

tjfulle commented:
Looks good to me.

haampie commented:
FWIW, `SPACK_DISABLE_LOCAL_CONFIG` is now a boolean not based on value but on whether it's set or not. The docs and your proposal say otherwise. I'm still a fan of setting / unsetting cause otherwise we have to worry about what is truth-y and what is false-y. So, only the docs would have to change.

haampie commented:
I'm happy we did not end up in a situation where the effective misc cache directory was determined by `config:misc_cache`'s default config value `$misc_cache_path` which by default was initialized as (`$user_config_path` which by default is initialized as `~/.spack` or `$SPACK_USER_CONFIG_PATH` if set) or `$SPACK_MISC_CACHE_PATH` if that is set, or, since cli arguments usually override environment variables, overridden by the command line argument `-cconfig:misc_cache:[path]` value or the command line scope `--config-scope {path}` `{path}/config.yaml`'s `config:misc_cache` value, which may or may not use `$misc_cache_path`.

Although it would be very customizable!

tgamblin commented:
Ok I think we're in good shape now -- squashed to one commit and set to rebase.

tgamblin commented:
Thanks everyone!

haampie commented:
github actions doesn't want it to be merged, maybe just drop the pr?

```
=================================== FAILURES ===================================
________________________ test_list_format_version_json _________________________

    @pytest.mark.maybeslow
    def test_list_format_version_json():
        output = list('--format', 'version_json')
>       assert '  {"name": "cloverleaf3d",' in output
E       assert '  {"name": "cloverleaf3d",' in '[\n  {"name": "3dtk",\n   "latest_version": "1.2",\n   "versions": ["trunk", "1.2"],\n   "homepage": "http://slam6d.s...topcf",\n   "latest_version": "1.0.5",\n   "versions": ["1.0.5"],\n   "homepage": "https://cgit.freedesktop.org/x\n]\n'

lib/spack/spack/test/cmd/list.py:58: AssertionError
!!!!!!!!!!!!!!!!!!!! Interrupted: stopping after 1 failures !!!!!!!!!!!!!!!!!!!!
== 1 failed, 2429 passed, 7 skipped, 22 xfailed, 5 xpassed in 1106.24 seconds ==
```
same issue in another pr of mine

tgamblin commented:
It’s done y’all!

haampie commented:
yay

mathstuf commented:
So I tested this out while also linking `~/.spack -> /dev/null` to catch anything else. I end up with this error:

```
==> Error: [Errno 20] Not a directory: '/Users/svc-dashboard/.spack/bootstrap'
```

So there are still bits that want to touch `~/.spack` around.

tgamblin commented:
@mathstuf Did you use both SPACK_DISABLE_LOCAL_CONFIG and SPACK_USER_CACHE_PATH? If you don’t use the latter, caches like bootstrap store will still default to the home directory.

mathstuf commented:
Yep:

```yaml
        SPACK_USER_CACHE_PATH: "$GIT_CLONE_PATH/.gitlab/spack-cache"
        SPACK_DISABLE_LOCAL_CONFIG: "true"
```

haampie commented:
So, this is because `$user_config_path` is not magically gone if you use `SPACK_DISABLE_LOCAL_CONFIG` :upside_down_face: 

This means that you still have to override all config settings that use `$user_config_path` or use `SPACK_USER_CONFIG_PATH`.

I think the better solution now is to get rid of `SPACK_DISABLE_LOCAL_CONFIG`, it's mostly confusing.

Instead stick to `SPACK_USER_CONFIG_PATH`, `SPACK_SYSTEM_CONFIG_PATH` and `SPACK_USER_CACHE_PATH`.

tgamblin commented:
@mathstuf: can you send a debug trace for the failure you mentioned above? (run with `spack -d`)

Do you refer to `user_config_path` in your configuration?  Is that where it's coming from?

trws commented:
My guess would be it's this biting us: https://github.com/spack/spack/pull/26735/files#diff-24ac473c44f26728fd3da17d5dd741c4073fa0d8aec651971c2aaf7d17747c9fR7

That still needs to go somewhere, probably fall back to the next config scope like the rest of spack does?

mathstuf commented:
> @mathstuf: can you send a debug trace for the failure you mentioned above? (run with spack -d)

I'll try and get a run with this in.

> Do you refer to user_config_path in your configuration? Is that where it's coming from?

Not that I know of. The entirety of the Spack on the machine (besides the `~/.spack` tripwire) is in this script (I removed the noisy Linux setup parts):

<details><summary>Spack setup script</summary>

Environment:

```yaml
        SPACK_USER_CACHE_PATH: "$GIT_CLONE_PATH/.gitlab/spack-cache"
        SPACK_DISABLE_LOCAL_CONFIG: "true"
```

```sh
#!/bin/sh

set -e

readonly spackroot="$CI_PROJECT_DIR/.gitlab/spack"

# Clone Spack itself.
git clone --depth 1 https://github.com/spack/spack "$spackroot/src"
export SPACK_ROOT="$spackroot/src"
export PATH="$PATH:$SPACK_ROOT/bin"

# Search for the compiler.
spack compiler find

# Clone mochi package repository.
git clone --depth 1 https://github.com/mochi-hpc/mochi-spack-packages "$spackroot/mochi"

# Add mochi repo to Spack.
spack repo add "$spackroot/mochi"

# Inject CI tools to avoid builds.
mkdir -p "$spackroot/config"
cat > "$spackroot/config/packages.yaml" <<EOF
packages:
  all:
    target: [$( uname -m )]
  cmake:
    buildable: false
    externals:
    - spec: cmake@$( cmake --version | head -n1 | cut -d' ' -f3 )
      prefix: $CI_PROJECT_DIR/.gitlab/cmake
EOF

readonly spec="mochi-thallium@main^libfabric fabrics=tcp,rxm"

# Install thallium.
spack -C "$spackroot/config" install $spec

# Clean up the directory.
spack -C "$spackroot/config" clean
```

</details>

tgamblin commented:
@mathstuf: I think I've fixed this over in #27022 -- gripping around for hard-coded `.spack` references.  I assumed the variables in `spack.paths` were actually being used as defaults here and that was clearly wrong.
